tasks/Functions-admin/manageAllowlist.js

```diff
@@ -43,14 +56,188 @@ async function addOrRemove(action, taskArgs) {
+  const titles = allowListLines[0]
+  const lines = allowListLines.slice(1)
+
+  // allUsers maintains the list of users in sorted order
+  const allUsers = []
+  const validUsers = []
+  const invalidUsers = []
+
+  // Split address from email and place into an object
+  for (const line of lines) {
+    const columns = line.split(',')
+
+    // Remove extraneous spaces from user's typed entries
+    const user = {
+      approved: columns[0],
+      walletAddress: columns[1].replace(/\s/g, ''),
+      email: columns[2].replace(/\s/g, ''),
+      agreedToTOS: columns[3],
+      eventCode: columns[4].replace(/\s/g, ''),
+      submittedAt: columns[5],
+      token: columns[6],
+      notes: columns[7],
+    }
+    // Maintain a list of all users
+    allUsers.push(user)
+    // Check that user is valid
+    if (!ethers.utils.isAddress(user.walletAddress)) {
+      user.notes = 'INVALID WALLET ADDRESS ' + user.notes
+      invalidUsers.push(user)
+      continue
+    }
+    if (user.agreedToTOS.toLowerCase() !== 'true') {
+      user.notes = 'DID NOT AGREE TO TOS' + user.notes
+      invalidUsers.push(user)
+      continue
+    }
+    if (
+      user.eventCode.toLowerCase() === '' ||
+      !requiredEventCodes.toLowerCase().split(',').includes(user.eventCode.toLowerCase())
+    ) {
+      user.notes = 'INVALID EVENT CODE' + user.notes
+      invalidUsers.push(user)
+      continue
+    }
+    // If user is valid, add them to validUsers
+    validUsers.push(user)
+  }
+
+  return {
+    titles,
+    allUsers,
+    validUsers,
+    invalidUsers,
+  }
+}
+
+const filterAddressesToAdd = (currentAllowlist, validUsers) => {
+  const currentAllowlistMap = {}
+
+  currentAllowlist.forEach((address) => {
+    currentAllowlistMap[address.toLowerCase()] = true
+  })
+
+  const addressesToAdd = []
+
+  for (const user of validUsers) {
+    if (!currentAllowlistMap[user.walletAddress.toLowerCase()]) {
+      addressesToAdd.push(user.walletAddress)
+    }
+  }
+
+  return addressesToAdd
+}
+
+const generateUpdatedAllowlistCsv = (updatedOnChainAllowlist, allowlistData) => {
+  const onChainAllowlistMap = {}
+
+  updatedOnChainAllowlist.forEach((address) => {
+    onChainAllowlistMap[address.toLowerCase()] = true
+  })
+
+  const lines = [ allowlistData.titles ]
+
+  for (let i = 0; i < allowlistData.allUsers.length; i++) {
+    if (onChainAllowlistMap[allowlistData.allUsers[i].walletAddress.toLowerCase()]) {
+      allowlistData.allUsers[i].approved = 'TRUE'
+    } else {
+      allowlistData.allUsers[i].approved = 'FALSE'
+    }
+
+    const {
+      approved,
+      walletAddress,
+      email,
+      agreedToTOS,
+      eventCode,
+      submittedAt,
+      token,
+      notes
+    } = allowlistData.allUsers[i]
+
+    lines.push(`${approved},${walletAddress},${email},${agreedToTOS},${eventCode},${submittedAt},${token},${notes}`)
+  }
+
+  const updatedCsvText = lines.join('\n')
+  fs.writeFileSync('./updatedAllowlist.csv', updatedCsvText)
```

Comment:
Why do we need this?

Reply:
We don't need it, but it makes it visually very easy to open this new file in Excel to check the diff

Ideally, you download the allowlist.csv Google sheet, run the command and paste updatedAllowlist.csv back into the Google sheet.

Reply:
It will automatically update the `Allowed/Approved` column, and if the address, event code or TOS agreement is invalid, it will add a message indicating this to the `Notes` column.  I added a description for this functionality to the README.